### PR TITLE
Optimizer should classify assignments to PseudoVariables as indirect assignments

### PR DESCRIPTION
**Pull request by VSadov**
In a rare case incorrect classification may lead the optimizer to believe it could get the RHS  of an assignment from the stack where in fact the stack will contain an LHS reference on the top.

Note that the condition when this would matter is extremely rare.
-- We need to have an LHS of an indirect assignment that is completely symbolic and not taking any arguments, indices, receivers, etc.. (otherwise we would arguments on the stack would prevent the bug from happening)
PseudoVariable is one of such variables, (it is used only by EE at debug time)

-- We need two subsequent assignments of the same local targeting incorrectly classified variables.

These conditions are not easily met, however we had a case where a pattern emitted by Deconstruction at debug time meeting these conditions and causing incorrect IL to happen.
It is likely that the problem is temporary and will disappear after recent changes in Deconstruction codegen. 

However we still should not have incorrect code waiting for something to expose the bug.

In this change:
1) Changed classification of PseudoVariable assignments (in both C# and VB)
2) Made classification exhaustive, so that asserts would be triggered if we need to analyse new assignment target.

NOTE: I tried, but could not create a testcase that would not involve Deconstruction pattern (that is going away).
It does not mean it is impossible, I just could not figure how.

Fixes:#15425

### CODE REVIEW

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - please review

**Comment by VSadov** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest  please
@dotnet-bot test windows_release_vs-integration_prtest  please

**Comment by VSadov** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by VSadov** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler  - ping

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs [hunk 1]
```diff
@@ -1245,6 +1245,11 @@ internal BoundExpression NullOrDefault(TypeSymbol typeSymbol)
 #endif
             )
         {
+            if (refKind == RefKind.Out)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: >if (refKind == RefKind.Out) [](start = 12, length = 27)

Should we even allow RefKind.Out as an input?

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs (hunk 1 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs]: As I found it is very convenient when creating a temp for a parameter - you just pass the ref kind of the parameter. 
I did not want to change all that code, just intercept "Out" in the helper.

File: src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb [hunk 2]
```diff
@@ -613,26 +613,45 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGen
...
-            Private Shared Function IsByRefLocalOrParameter(node As BoundExpression) As Boolean
+            Private Shared Function IsByRefVariable(node As BoundExpression) As Boolean
                 Select Case node.Kind
                     Case BoundKind.Parameter
                         Return DirectCast(node, BoundParameter).ParameterSymbol.IsByRef
                     Case BoundKind.Local
                         Return DirectCast(node, BoundLocal).LocalSymbol.IsByRef
-                    Case Else
+                    Case BoundKind.Call
+                        Return DirectCast(node, BoundCall).Method.ReturnsByRef
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb]: Are we testing the case where `Method.ReturnsByRef = True`?

File: src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb [hunk 3]
```diff
@@ -613,26 +613,45 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGen
...
             ''' <summary>
-            ''' indirect assignment is assignment to a value referenced indirectly
-            ''' it may only happen if lhs is a reference (must be a parameter or a local)
-            '''       1) lhs is a reference (must be a parameter or a local)
-            '''       2) it is not a ref/out assignment where the reference itself would be assigned
+            ''' VB uses a special node to assign references.
+            ''' BoundAssignment is used only to assign values.
+            ''' therefore an indirect assignment may only happen if lhs is a reference
             ''' </summary>
             Friend Shared Function IsIndirectAssignment(node As BoundAssignmentOperator) As Boolean
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb]: `Private`?

**Comment by cston** [comment]
LGTM

File: src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb (hunk 3 above)
Comment (VSadov) [inline src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb]: Fixed

File: src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb (hunk 2 above)
Comment (VSadov) [inline src/Compilers/VisualBasic/Portable/CodeGen/Optimizer/StackScheduler.Analyzer.vb]: Will add a test for both VB and C#. We did have a bug here in VB, that this change also fixed.

**Comment by VSadov** [comment]
test this please
No tests runnig after 1hr waiting

**Pull request merged**